### PR TITLE
Feature/questions

### PR DESCRIPTION
**Pull request by tubhutani13**

### CODE REVIEW

File: db/migrate/20230120052312_create_questions.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class CreateQuestions < ActiveRecord::Migration[7.0]
+  def change
+    create_table :questions do |t|
+      t.string :title
+      t.string :published_token
```
Comment (shubhamjuneja1) [inline db/migrate/20230120052312_create_questions.rb:5]: rename to permalink

File: lib/tasks/set_token.rake [hunk 2]
```diff
@@ -0,0 +1,12 @@
+namespace :question do
+    desc "Question"
+    task create_token: :environment do
+        Question.all.each do |question|
```
Comment (shubhamjuneja1) [inline lib/tasks/set_token.rake]: can't this be just
Question.where(published_token: nil)

File: lib/tasks/set_token.rake [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+            if question.published_token.nil?
+                question.generate_token(:published_token)
+                question.save(:validate => false)
```
Comment (shubhamjuneja1) [inline lib/tasks/set_token.rake]: why skipping validations?

File: lib/tasks/set_token.rake (hunk 2 above)
Comment (shubhamjuneja1) [reply lib/tasks/set_token.rake]: check spaces as well

File: app/models/user.rb [hunk 4]
```diff
@@ -1,29 +1,29 @@
 class User < ApplicationRecord
+  include ::TokenHandler
   enum role: {
          "admin" => 1,
          "user" => 0,
        }
 
+  has_many :questions, dependent: :nullify
```
Comment (shubhamjuneja1) [inline app/models/user.rb]: dependent: :restrict_with_error

File: app/models/concerns/token_handler.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+module TokenHandler
+  def generate_token(column)
```
Comment (shubhamjuneja1) [inline app/models/concerns/token_handler.rb:2]: should just return token not update column

File: app/models/question.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
+class Question < ApplicationRecord
+  include ::TokenHandler
+
+  belongs_to :user
+
+  before_create -> { generate_token(:published_token) }
```
Comment (shubhamjuneja1) [inline app/models/question.rb]: validate unique token

File: app/models/question.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  before_save :ensure_published_question_cannot_be_drafted
+
+  with_options if: :published? do
+    validates :title, presence: true, uniqueness: true
+    validates :content, presence: true, length: { minimum: 15 }
+    validates :topic_list, presence: true
+  end
+
+  has_one_attached :pdf_attachment
+  has_rich_text :content
```
Comment (shubhamjuneja1) [inline app/models/question.rb:20]: keep things at one place

validations
associations

methods
private methods

File: app/models/question.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+  before_save :ensure_published_question_cannot_be_drafted
+
+  with_options if: :published? do
+    validates :title, presence: true, uniqueness: true
+    validates :content, presence: true, length: { minimum: 15 }
+    validates :topic_list, presence: true
+  end
+
+  has_one_attached :pdf_attachment
```
Comment (shubhamjuneja1) [inline app/models/question.rb:19]: it should not be pdf_attachment, as filr can be of any type

File: app/models/question.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  acts_as_taggable_on :topics
+
+  private def ensure_published_question_cannot_be_drafted
+    if changes[:published] == [true, false]
+      errors.add(:base, "Published question cannot be Drafted")
+      throw :abort
+    end
+  end
+
+  def to_param
```
Comment (shubhamjuneja1) [inline app/models/question.rb:30]: why?

File: app/models/question.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    published_token
+  end
+  
+  def published?
+    return !(self.published_at.nil?)
```
Comment (shubhamjuneja1) [inline app/models/question.rb]: can't we simply write
published_at.present?

File: app/models/question.rb [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  def publish_question(published)
+    check_publish(published)
```
Comment (shubhamjuneja1) [inline app/models/question.rb:39]: why are we passing published as an argument here?
check in controller if this is true and call publish

File: app/models/question.rb [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    save
+  end
+
+  def update_question(params, published)
+    check_publish(published)
+    update(params)
```
Comment (shubhamjuneja1) [inline app/models/question.rb:45]: same as above

File: app/controllers/questions_controller.rb [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    @question = Question.new
+  end
+
+  def edit
+  end
+
+  def create
+    @question = current_user.questions.build(question_params)
+    respond_to do |format|
+      if @question.publish_question(params[:publish])
```
Comment (shubhamjuneja1) [inline app/controllers/questions_controller.rb:14]: create should not publish directly and even if that's required use attr_accessor isntead of passing params
let's discuss

File: app/controllers/questions_controller.rb [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        if params[:draft]
```
Comment (shubhamjuneja1) [inline app/controllers/questions_controller.rb:15]: should be automatically handle using states

File: app/controllers/questions_controller.rb [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+      else
+        flash[:error] = "Ooooppss, something went wrong!"
+        format.html { render :new, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  def update
+    if @question.update_question(question_params, params[:publish])
+      if params[:draft]
```
Comment (shubhamjuneja1) [inline app/controllers/questions_controller.rb:30]: same

File: app/controllers/questions_controller.rb [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+      else
+        redirect_to question_path, notice: t("Question updated successfully")
+      end
+    else
+      render :edit, status: :unprocessable_entity
+    end
+  end
+
+  def destroy
+    @question.destroy
```
Comment (shubhamjuneja1) [inline app/controllers/questions_controller.rb]: what if this fails?

**Review by shubhamjuneja1 — CHANGES_REQUESTED** [review]

File: app/models/question.rb (hunk 8 above)
Comment (tubhutani13) [reply app/models/question.rb:19]: it was written that file can be pdf only

File: app/models/question.rb (hunk 6 above)
Comment (tubhutani13) [reply app/models/question.rb]: already handled in generate token
discuss please

File: app/models/question.rb (hunk 9 above)
Comment (tubhutani13) [reply app/models/question.rb:30]: so that anywhere when link or url is used it will check the permalink instead of id

**Pull request closed without merging**